File: netlify/functions/lib/money.ts
export function parseMoneyCents(value: string): number {
  let raw = value.trim().replace("R$", "").replace(/ /g, "");
  if (!raw) throw new Error("Informe um valor.");

  if (raw.includes(",")) {
    raw = raw.replace(/\./g, "").replace(",", ".");
  } else if (raw.includes(".")) {
    const lastGroup = raw.split(".").pop()!;
    if (lastGroup.length === 3) raw = raw.replace(/\./g, "");
  }

  const n = parseFloat(raw);
  if (isNaN(n)) throw new Error(`Valor invalido: ${value}`);
  return Math.round(n * 100);
}

export function formatCents(cents: number): string {
  const sign = cents < 0 ? "-" : "";
  const abs = Math.abs(cents);
  const reais = Math.floor(abs / 100);
  const centavos = abs % 100;
  const inteiro = reais.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ".");
  return `${sign}R$ ${inteiro},${centavos.toString().padStart(2, "0")}`;
}

export function splitAmount(totalCents: number, parts: number): number[] {
  const base = Math.floor(totalCents / parts);
  const remainder = totalCents % parts;
  const amounts = new Array<number>(parts).fill(base);
  amounts[amounts.length - 1] += remainder;
  return amounts;
}

export function serializeMoneyMap(
  values: Record<string, number>
): Record<string, { cents: number; label: string }> {
  return Object.fromEntries(
    Object.entries(values).map(([k, v]) => [k, { cents: v, label: formatCents(v) }])
  );
}

File: finance_app/domain/money.py
from decimal import Decimal, InvalidOperation, ROUND_HALF_UP


def parse_money_to_cents(value: str) -> int:
    raw = value.strip().replace("R$", "").replace(" ", "")
    if not raw:
        raise ValueError("Informe um valor.")

    if "," in raw:
        raw = raw.replace(".", "").replace(",", ".")
    elif "." in raw:
        last_group = raw.rsplit(".", 1)[-1]
        if len(last_group) == 3:
            raw = raw.replace(".", "")

    try:
        decimal_value = Decimal(raw)
    except InvalidOperation as exc:
        raise ValueError(f"Valor invalido: {value}") from exc

    cents = (decimal_value * 100).quantize(Decimal("1"), rounding=ROUND_HALF_UP)
    return int(cents)


def format_cents(cents: int) -> str:
    sign = "-" if cents < 0 else ""
    cents = abs(cents)
    reais, centavos = divmod(cents, 100)
    inteiro = f"{reais:,}".replace(",", ".")
    return f"{sign}R$ {inteiro},{centavos:02d}"


def split_amount(total_cents: int, parts: int) -> list[int]:
    if parts <= 0:
        raise ValueError("A quantidade de parcelas deve ser maior que zero.")

    base, remainder = divmod(total_cents, parts)
    amounts = [base for _ in range(parts)]
    amounts[-1] += remainder
    return amounts
